src/r3d.imageio/r3dinput.cpp

```diff
@@ -0,0 +1,441 @@
+unsigned char*
+AlignedMalloc(size_t& sizeNeeded)
+{
+    // alloc 15 bytes more to make sure we can align the buffer in case it isn't
+    unsigned char* buffer = (unsigned char*)malloc(sizeNeeded + 15U);
+
+    if (!buffer)
+        return nullptr;
+
+    sizeNeeded = 0U;
+
+    // cast to a 32-bit or 64-bit (depending on platform) integer so we can do the math
+    uintptr_t ptr = (uintptr_t)buffer;
+
+    // check if it's already aligned, if it is we're done
+    if ((ptr % 16U) == 0U)
+        return buffer;
+
+    // calculate how many bytes we need
+    sizeNeeded = 16U - (ptr % 16U);
+
+    return buffer + sizeNeeded;
+}
+
+OIIO_PLUGIN_NAMESPACE_BEGIN
+
+#define DBG if (1)
+
+class R3dInput final : public ImageInput {
+public:
+    R3dInput()
+    {
+        initialize();
+        reset();
+    }
+    ~R3dInput() override
+    {
+        close();
+        terminate();
+    }
+    const char* format_name(void) const override { return "r3d"; }
+    int supports(string_view feature) const override
+    {
+        return (feature == "ioproxy");
+    }
+    bool open(const std::string& name, ImageSpec& spec) override;
+    bool open(const std::string& name, ImageSpec& spec,
+              const ImageSpec& config) override;
+    bool seek_subimage(int subimage, int miplevel) override;
+    bool read_native_scanline(int subimage, int miplevel, int y, int z,
+                              void* data) override;
+    void read_frame(int pos);
+    int current_subimage(void) const override
+    {
+        lock_guard lock(*this);
+        return m_subimage;
+    }
+    bool close() override;
+
+    const std::string& filename() const { return m_filename; }
+
+    bool seek(int pos);
+    double fps() const;
+    int64_t time_stamp(int pos) const;
+
+private:
+    std::string m_filename;
+    std::unique_ptr<ImageSpec> m_config;  // Saved copy of configuration spec
+    R3DSDK::Clip* m_clip;
+    R3DSDK::VideoDecodeJob m_job;
+    unsigned char* m_image_buffer;
+    size_t m_adjusted;
+    int m_frames;
+    int m_channels;
+    float m_fps;
+    int m_subimage;
+    int64_t m_nsubimages;
+    int m_last_search_pos;
+    int m_last_decoded_pos;
+    bool m_read_frame;
+    int m_next_scanline;
+    // std::vector<uint16_t> m_pixels;
+
+    void initialize();
+    void reset()
+    {
+        DBG std::cout << "R3dInput::reset()\n";
+
+        ioproxy_clear();
+        m_config.reset();
+        m_clip             = nullptr;
+        m_next_scanline    = 0;
+        m_read_frame       = false;
+        m_subimage         = 0;
+        m_last_search_pos  = 0;
+        m_last_decoded_pos = 0;
+        m_image_buffer     = nullptr;
+        m_adjusted         = 0;
+    }
+
+    void close_file() { reset(); }
+    void terminate();
+};
+
+// Export version number and create function symbols
+OIIO_PLUGIN_EXPORTS_BEGIN
+
+OIIO_EXPORT int r3d_imageio_version = OIIO_PLUGIN_VERSION;
+
+OIIO_EXPORT const char*
+r3d_imageio_library_version()
+{
+    return "R3D";
```

Comment:
This should say a version number, if it's at all possible to know.

Reply:
I have observed this output during run, so it can be obtained and used during build phase as well.

R3dInput::initialize()
SDK VERSION: R3DAPI 8.5.1-53601 (20240308 Lx64S) R3DSDK 8.5.1-53601 (20240308 Lx64D C3B1) [/opt/R3DSDKv8_5_1/Redistributable/linux/]